File: compiler/nodes/patterns/tuple.go
package patterns

import (
	"fmt"

	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type TuplePatternNode struct {
	Elements []database.Node
	Facts    *database.Facts

	matching database.Node

	elementTemporaries []database.Node
}

func (node *TuplePatternNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseTuplePattern(parser *syntax.Parser) (*TuplePatternNode, *syntax.Error) {
	span := parser.Spanned()

	many, err := syntax.ParseMany(parser, 1, ParsePatternElement, func(parser *syntax.Parser) (*struct{}, *syntax.Error) {
		_, err := parser.Token("TupleOperator")
		if err != nil {
			return nil, err
		}

		parser.ConsumeLineBreaks()

		return &struct{}{}, nil
	})
	if err != nil {
		return nil, err
	}

	if len(many) == 1 {
		_, err = parser.Token("TupleOperator")
		if err != nil {
			return nil, err
		}
	} else {
		_, _, err := syntax.ParseOptional(parser, func(parser *syntax.Parser) (struct{}, *syntax.Error) {
			_, err := parser.Token("TupleOperator")
			if err != nil {
				return struct{}{}, err
			}

			return struct{}{}, nil
		})
		if err != nil {
			return nil, err
		}
	}

	elements := make([]database.Node, 0, len(many))
	for _, item := range many {
		elements = append(elements, item.Value)
	}

	return &TuplePatternNode{
		Elements: elements,
		Facts:    database.NewFacts(span()),
	}, nil
}

func (node *TuplePatternNode) Visit(visitor *visit.Visitor) {
	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node

	node.elementTemporaries = make([]database.Node, 0, len(node.Elements))
	for _, element := range node.Elements {
		node.elementTemporaries = append(node.elementTemporaries, visitor.VisitMatching(element))
		visitor.Db.Graph.Edge(element, node, "element")
	}

	visitor.Constraint(typecheck.NewTypeConstraint(node, typecheck.TupleType(node.elementTemporaries)))
}

func (node *TuplePatternNode) Codegen(c *codegen.Codegen) error {
	if len(node.elementTemporaries) != len(node.Elements) {
		return c.Error(node)
	}

	span := database.GetSpanFact(node)

	for index, element := range node.elementTemporaries {
		c.WriteString(span, " && ((")
		c.WriteNode(span, element)
		c.WriteString(span, " = ")
		c.WriteNode(span, node.matching)
		c.WriteString(span, "[")
		c.WriteString(span, fmt.Sprintf("%d", index))
		c.WriteString(span, "]) || true)")

		if err := c.Write(node.Elements[index]); err != nil {
			return err
		}
	}

	return nil
}

func (node *TuplePatternNode) EachTemporary(f func(database.Node)) {
	for _, temporary := range node.elementTemporaries {
		f(temporary)
	}

	for _, element := range node.Elements {
		EachTemporary(element, f)
	}
}

File: compiler/src/nodes/patterns/tuple.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Node, NodeRef},
    nodes::{Matching, Temporaries, parse_pattern_element, visit_pattern},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{MatchPathSegment, Visit, Visitor},
};

#[derive(Debug)]
pub struct TuplePatternNode {
    pub elements: Vec<NodeRef>,
}

impl Node for TuplePatternNode {}

pub fn parse_tuple_pattern(parser: &mut Parser<'_>) -> Result<TuplePatternNode, ParseError> {
    let elements = parser
        .parse_many(1, parse_pattern_element, |parser| {
            parser.token(TokenKind::TupleOperator)?;
            parser.consume_line_breaks();
            Ok(())
        })?
        .into_iter()
        .map(|(node, _)| node)
        .collect::<Vec<_>>();

    if elements.len() == 1 {
        parser.token(TokenKind::TupleOperator)?;
    } else {
        let _ = parser.parse_optional(|parser| {
            parser.token(TokenKind::TupleOperator)?;
            Ok(())
        })?;
    }

    Ok(TuplePatternNode { elements })
}

impl Visit for TuplePatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_pattern(node, visitor, None);

        let element_temporaries = self
            .elements
            .iter()
            .enumerate()
            .map(|(index, element)| {
                let temporary = visitor.visit_matching(
                    element,
                    Some(MatchPathSegment::TupleElement(index, self.elements.len())),
                );

                visitor.edge(element, node, "element");

                temporary
            })
            .collect::<Vec<_>>();

        visitor.constraint(TypeConstraint::new(
            node.clone(),
            visitor.tuple_type(element_temporaries.clone()),
        ));

        visitor.insert(
            node,
            Temporaries {
                has: element_temporaries,
                inherit: self.elements.clone(),
            },
        )
    }
}

impl Codegen for TuplePatternNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let Temporaries {
            has: element_temporaries,
            ..
        } = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        let Matching(matching) = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        for (index, (element, temporary)) in
            self.elements.iter().zip(element_temporaries).enumerate()
        {
            ctx.condition(ir::Condition::Initialize {
                variable: temporary,
                node: None,
                value: ir::Value::TupleElement {
                    input: matching.clone(),
                    index,
                },
                mutable: false,
            });

            ctx.codegen(element)?;
        }

        Ok(())
    }
}
